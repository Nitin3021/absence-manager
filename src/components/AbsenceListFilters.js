import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, 
         setStartDate, 
         setEndDate, 
         sortByCreatedDate, 
         sortByStartDate, 
         sortByEndDate } from '../actions/filters';

export class AbsenceListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if (e.target.value === 'createdAt') {
            this.props.sortByCreatedDate();
        } else if (e.target.value === 'startDate') {
            this.props.sortByStartDate();
        } else if (e.target.value === 'endDate') {
            this.props.sortByEndDate();
        }
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            placeholder="Search absence type"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="createdAt">Created On</option>
                            <option value="startDate">Start Date</option>
                            <option value="endDate">End Date</option>
                        </select>
                    </div>
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByCreatedDate: () => dispatch(sortByCreatedDate()),
    sortByStartDate: () => dispatch(sortByStartDate()),
    sortByEndDate: () => dispatch(sortByEndDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(AbsenceListFilters);