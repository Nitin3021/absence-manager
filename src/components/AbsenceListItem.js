import React from 'react';
import moment from 'moment';

const AbsenceListItem = ({ absences }) => {
    return absences.map((absence) => {
        const { key, name, type, createdAt, startDate, endDate, admitterNote } = absence;
        return (
            <tr key={key}>
                <td>{name}</td>
                <td>{type}</td>
                <td>{moment(createdAt).format('DD.MMM.YYYY')}</td>
                <td>{moment(startDate).format('DD.MMM.YYYY')}</td>
                <td>{moment(endDate).format('DD.MMM.YYYY')}</td>
                <td>{admitterNote}</td>
            </tr>
        )
    })
}

export default AbsenceListItem;