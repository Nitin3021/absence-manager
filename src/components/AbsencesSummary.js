import React from 'react';
import { connect } from 'react-redux';
import selectAbsences from '../selectors/absences';

export const AbsencesSummary = (props) => {
    const absenceWord = props.absenceCount === 1 ? 'Absence' : 'Absences';
    const hiddenAbsenceCountWord = props.hiddenAbsenceCount === 1 ? 'Absence' : 'Absences';

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing total {props.absenceCount} {absenceWord}</h1>
                {
                    props.hiddenAbsenceCount === 0 ? (
                        <p>No filtered absence!</p>
                    ) : (
                        <p>Filtered {props.hiddenAbsenceCount} {hiddenAbsenceCountWord} </p>
                    )
                }
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleAbsences = selectAbsences(state.absences, state.filters)

    return {
        absenceCount: visibleAbsences.length,
        hiddenAbsenceCount: state.absences.length - visibleAbsences.length
    }
}

export default connect(mapStateToProps)(AbsencesSummary);