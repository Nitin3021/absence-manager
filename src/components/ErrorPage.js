import React from 'react';

const ErrorPage = () => (
    <div className="error-layout">
        <h1 className="error-text">An error encountered while fetching, please try again later!</h1>
    </div>
);

export default ErrorPage;