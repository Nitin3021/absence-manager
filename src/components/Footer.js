import React from 'react';

const Footer = () => (
    <div className="footer">
        <h1>Â© NP</h1>
    </div>
)

export default Footer;