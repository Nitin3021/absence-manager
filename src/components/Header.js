import React from 'react';

const Header = () => (
    <header>
        <h1>Absence Manager</h1>
    </header>
);

export default Header;