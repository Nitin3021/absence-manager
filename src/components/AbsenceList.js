import React from 'react';
import { connect } from 'react-redux';
import AbsenceListItem from './AbsenceListItem';

export const AbsenceList = (props) => (
    <div>
        <div>
            {
                props.absences.length === 0 ? (
                    <div>
                        <span>No Absences!</span>
                    </div>
                ) : (
                    props.absences.map((absence) => {
                        return <AbsenceListItem key={absence.key} {...absence} />;
                    })
                )
            }
        </div>
    </div>
)

const mapStateToProps = (state) => ({
    absences: state.absences
})

export default connect(mapStateToProps)(AbsenceList);