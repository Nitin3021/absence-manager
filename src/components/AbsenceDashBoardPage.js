import React from 'react';
import AbsenceList from './AbsenceList';
import AbsenceListFilters from './AbsenceListFilters';
import AbsencesSummary from './AbsencesSummary';

const AbsenceDashBoardPage = () => (
    <div>
        <AbsencesSummary />
        <AbsenceListFilters />
        <AbsenceList />
    </div>
);

export default AbsenceDashBoardPage;