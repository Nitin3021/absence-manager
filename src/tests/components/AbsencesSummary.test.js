import React from 'react';
import { shallow } from 'enzyme';
import { AbsencesSummary } from '../../components/AbsencesSummary';

test('should correctly render AbsencesSummary with 1 absence', () => {
    const wrapper = shallow(<AbsencesSummary absenceCount={1} absenceHiddenCount={41} />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render AbsencesSummary with multiple absences', () => {
    const wrapper = shallow(<AbsencesSummary absenceCount={21} absenceHiddenCount={21} />);
    expect(wrapper).toMatchSnapshot();
});