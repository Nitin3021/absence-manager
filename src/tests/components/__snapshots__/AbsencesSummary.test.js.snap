// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render AbsencesSummary with 1 absence 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing total 
      1
       
      Absence
    </h1>
    <p
      className="page-header__filtered-absence"
    >
      Filtered 
       
      Absences
    </p>
  </div>
</div>
`;

exports[`should correctly render AbsencesSummary with multiple absences 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing total 
      21
       
      Absences
    </h1>
    <p
      className="page-header__filtered-absence"
    >
      Filtered 
       
      Absences
    </p>
  </div>
</div>
`;
